android-components/components/service/nimbus/src/main/java/mozilla/components/service/nimbus/messaging/NimbusMessagingStorage.kt

```diff
@@ -272,6 +250,63 @@ class NimbusMessagingStorage(
         return nextMessage
     }
 
+    /**
+     * Returns a pair of uuid and valid action for the provided [message].
+     *
+     * The message's action-params are appended as query parameters to the action URI,
+     * URI encoding the values as it goes.
+     *
+     * Uses Nimbus' targeting attributes to do basic string interpolation.
+     *
+     * e.g.
+     * `https://example.com/{locale}/whats-new.html?version={app_version}`
+     *
+     * If the string `{uuid}` is detected in the [message]'s action, then it is
+     * replaced with a random UUID. This is returned as the first value of the returned
+     * [Pair].
+     *
+     * The fully resolved (with all substitutions) action is returned as the second value
+     * of the [Pair].
+     */
+    fun generateUuidAndFormatAction(message: Message): Pair<String?, String> =
```

Comment:
nit: we could use a `typealias` for the strings used in the pair to make this a little more self describing, e.g. 
```kotlin
typealias Uuid = String
typealias Action = String
```

could be used in the fun below too